### PR TITLE
JENKINS-40912 list un/stashing files

### PR DESCRIPTION
**Pull request by joseblas**
See [JENKINS-40912](https://issues.jenkins-ci.org/browse/JENKINS-40912).

### Proposed changelog entries

* Entry 1: Adding a new method for archive and untar which have a Consumer<String> which will be run when any file get archived/untared.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [X] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@jtnord 
@dwnusbaum 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [x] There are at least 2 approvals for the pull request and no outstanding requests for change
- [x]  Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [x] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [x] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/test/java/hudson/FilePathTest.java
Comment (olamy) [inline core/src/test/java/hudson/FilePathTest.java:376]: would be better to have a real test on this rather than printing log.
maybe some assert on the file entries.

**Review by olamy — CHANGES_REQUESTED** [review]
sounds good. but need more testing. Including test on the previous `archive` method without the new arguments.

File: core/src/test/java/hudson/FilePathTest.java
Comment (joseblas) [reply core/src/test/java/hudson/FilePathTest.java]: also `archive` is used by tar which is widely used in the tests.

**Review by olamy — APPROVED** [review]

File: core/src/main/java/hudson/FilePath.java
Comment (jtnord) [inline core/src/main/java/hudson/FilePath.java:78]: unused import?

File: core/src/main/java/hudson/FilePath.java
Comment (jtnord) [inline core/src/main/java/hudson/FilePath.java:505]: should this not be a static class?

File: core/src/main/java/hudson/FilePath.java
Comment (jtnord) [inline core/src/main/java/hudson/FilePath.java:592]: does this not produce warnings at runtime about the serialization of this Lambda?

File: core/src/main/java/hudson/FilePath.java
Comment (jtnord) [inline core/src/main/java/hudson/FilePath.java:2676]: why dot pass the consume in here directly - so that you can get progress as it happens as opposed to after the fact (which is unhelpfull in the case something goes wrong)?

File: core/src/main/java/hudson/FilePath.java
Comment (jtnord) [inline core/src/main/java/hudson/FilePath.java:629]: if you want logging here - then I would assume that you want logging as it happens not after the fact (if you want after the fact you could as easily `ls` or the equivallent?

File: core/src/main/java/hudson/FilePath.java
Comment (jtnord) [inline core/src/main/java/hudson/FilePath.java:525]: this looks like it is called *after* the fact.  if somone wants logging of the operations then surely the want them as it happens.
for example if it takes 1 minute to archive/unarchive something then I would as a suser expect to see what was beeing archvied as it happened not nothing for 1 minute and then a sprew of everything.

File: core/src/main/java/hudson/util/io/Archiver.java
Comment (jtnord) [inline core/src/main/java/hudson/util/io/Archiver.java:50]: not needed.  should be the job of any `Consumer` to actually record this if it is wanted.

File: core/src/main/java/hudson/util/io/TarArchiver.java
Comment (jtnord) [inline core/src/main/java/hudson/util/io/TarArchiver.java:77]: a job for the `Consumer`?

**Review by jtnord — COMMENTED** [review]
IIUC the requirement it should be the job of any consumer to actually record and deal with the archived/unarchived files *as they occur* not be provided a list of them after the fact (which if was the case should just have a return that was List<String>.

thus I think this implementation is not correct, as the provided `Consumer` gets called at the end of the operations not during them.

I would also like to see a test of this over a remoting `Channel`

File: core/src/main/java/hudson/FilePath.java
Comment (joseblas) [reply core/src/main/java/hudson/FilePath.java:504]: I wouldn't change it

File: core/src/main/java/hudson/FilePath.java
Comment (joseblas) [reply core/src/main/java/hudson/FilePath.java]: without that Serializable it fails, and had to be added to remote calls (FilePathSecureTest)

Comment (jtnord) [reply core/src/main/java/hudson/FilePath.java]: I was more thinking that it would produce a warning like anonymous classes do. (I have not ever tried to send a Lambda over remoting)  https://www.jenkins.io/doc/developer/extensibility/serialization-of-anonymous-classes/

Comment (joseblas) [reply core/src/main/java/hudson/FilePath.java]: without the serialised interface you get this error https://ci.jenkins.io/blue/organizations/jenkins/Core%2Fjenkins/detail/PR-5061/5/tests

File: core/src/main/java/hudson/FilePath.java
Comment (jtnord) [inline core/src/main/java/hudson/FilePath.java:2707]: I think this is wrong - can you add a unit test for it?

if the tar entry is `foo/bar/wibble.txt` I think this will just pass `wibble.txt` to the consumer and not `foo/bar/wibble.txt`

Comment (jtnord) [reply core/src/main/java/hudson/FilePath.java:2707]:
```suggestion
                consumer.accept(e.getName());
```

will give you the relative path in the tar, or 

```java
                consumer.accept(f.toPath().relativise(baseDir.toPath()).toString());
```

may well give you the relative path (untested)

**Review by fcojfernandez — APPROVED** [review]

File: core/src/main/java/hudson/FilePath.java
Comment (MRamonLeon) [inline core/src/main/java/hudson/FilePath.java:491]: Could you complete the javadoc with the @params, @since an @throw sections please? It will help clients of this new method

File: core/src/main/java/hudson/util/io/TarArchiver.java
Comment (MRamonLeon) [inline core/src/main/java/hudson/util/io/TarArchiver.java:76]: shouldn't we enforce nothing wrong happens on the consumer? like try-catching it to avoid bad clients making the archiver to fail.

File: core/src/main/java/hudson/util/io/Archiver.java
Comment (MRamonLeon) [inline core/src/main/java/hudson/util/io/Archiver.java:48]: I suggest to use a name that makes us know the meaning of this consumer. A consumer to do what? do you know what I mean?

File: core/src/main/java/hudson/FilePath.java
Comment (MRamonLeon) [inline core/src/main/java/hudson/FilePath.java:508]: could you rename it to make it clear what this consumer function does? or when it is called to be more specific

File: core/src/main/java/hudson/FilePath.java
Comment (joseblas) [reply core/src/main/java/hudson/FilePath.java:491]: some of them are describe above, and I just copied and updated the javadoc from the overloaded method.

File: core/src/main/java/hudson/util/io/Archiver.java
Comment (joseblas) [reply core/src/main/java/hudson/util/io/Archiver.java:48]: is general purpose consumer, do not get wrong, but is a consumer to do whatever you want with the stashed files: list them, save them, print them...

File: core/src/main/java/hudson/FilePath.java
Comment (joseblas) [reply core/src/main/java/hudson/FilePath.java:508]: ditto

**Comment by timja** [comment]
Haven't really looked at the code but the changelog entry could do with improvements:

> Duplicate methods for archive/untar in order to pass a Consumer for processing the files involved

You're adding a feature, talk about that rather than that you duplicated methods

**Review by MRamonLeon — COMMENTED** [review]
Some more javadoc needed I believe. Instead of consumer I would use `afterUntarConsumer` or something like that, to denote when it's used. Although with a good javadoc is not so mandatory.

File: core/src/main/java/hudson/FilePath.java
Comment (MRamonLeon) [inline core/src/main/java/hudson/FilePath.java:532]: You can use a try-with-resource I believe

File: core/src/main/java/hudson/FilePath.java
Comment (MRamonLeon) [inline core/src/main/java/hudson/FilePath.java:501]: Missing the @throw section

**Review by MRamonLeon — APPROVED** [review]

**Comment by MRamonLeon** [comment]
IIUC @jtnord's comments have been addressed. Setting the ready for merge label to merge it after 24 hours.

**Review by jtnord — APPROVED** [review]

**Comment by dwnusbaum** [comment]
@joseblas Can you update https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/133 to use the current version of this PR to show the new methods in action?

**Comment by joseblas** [comment]
@dwnusbaum do yo want before merging this?

**Comment by MRamonLeon** [comment]
We may merge it in 24 hours if there is no negative feedback. Please see the merge process documentation for more information about the merge process

**Comment by dwnusbaum** [comment]
> do yo want before merging this?

That's up to the core maintainers. Personally I think it would make sense to demonstrate that the new methods actually work for the intended use case before merging.

File: core/src/main/java/hudson/FilePath.java
Comment (daniel-beck) [inline core/src/main/java/hudson/FilePath.java:496]:
```suggestion
     * @since TODO
```

**Review by daniel-beck — DISMISSED** [review]
I'm going to have to put on my security officer hat and block this for a bit until we've had an opportunity to evaluate this PR.

Nothing wrong with this change in particular, but might cause trouble with some other stuff we have going on.

Poke me in a week if I haven't followed up on this comment by then.

**Review by jeffret-b — APPROVED** [review]
I've reviewed this carefully and approve the changes. I haven't performed any manual testing, which isn't much of a possibility here, but have run and checked the tests.

Everything looks good. I'm especially glad to see the new tests, which provide sufficient coverage of the newly added functionality.

File: core/src/main/java/hudson/FilePath.java
Comment (jeffret-b) [reply core/src/main/java/hudson/FilePath.java:496]: This does need to be updated.

**Comment by jeffret-b** [comment]
@daniel-beck , I've completed my review here and am satisfied that this change is good and won't cause excessive problems with our other ongoing work.

**Comment by jglick** [comment]
> it would make sense to demonstrate that the new methods actually work for the intended use case

I suppose that would be https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/133? But that PR has merge conflicts. Need to have downstream plugin tests using a recent incremental build of core passing in order to validate that the core patch is correct and sufficient.

**Comment by alecharp** [comment]
@joseblas Could you review the latest merge conflicts?

**Comment by joseblas** [comment]
@alecharp I'm off for the next 2/3 weeks. Catch up later!

**Comment by basil** [comment]
Hi Jose, I took a look at this and have the following observations:

- The merge conflicts with recent security fixes are significant and require careful attention.
- The high-level problem is not 100% clear to me, but it is clear enough for me to recognize the need for a solution. :thumbsup:
- The design of this specific solution seems about 85% complete overall, but the API could possibly use some polishing. It is hard for me to provide specific details without seeing the intended consumer, but my gut feeling is that at least the naming of the arguments could be improved, if not the method signatures.
- The implementation also seems about 85% complete overall. Some minor aspects of code style and test implementation could be polished, but overall the implementation looks solid (though it may be subject to change given the design feedback from the previous point). I can provide suggestions, but it seems premature to do a detailed review of Java coding style before we finalize the design and API.

I recognize my involvement is quite late at this point, but I do think this PR is valuable and would be happy to shepherd it forward. However, the number of pending action items exceeded what I could complete myself in an hour or two, so I cannot carry this forward to completion alone. But I would be happy to work with you to drive this to the finish line.

From my perspective, the main thing I need as a reviewer is a complete understanding of the original problem and a link to a consumer demonstrating how this API will be used. You can contact me directly if desired. Once I have this, I can review the API and will likely have some feedback on the parameters and method signatures. Once we are done with that, we need to get the implementation up-to-date. Since merging with recent security changes is non-trivial, it may be easiest to simply reimplement the (by then, improved) design from scratch on top of recent trunk commits.

Thanks for everything you have done so far. For now, I am applying the `stalled` label since there has been no activity in a long time. If we move forward with this in the next few weeks, I will remove it; otherwise, I will probably apply the `proposed-for-close` label approximately a month after that.

**Pull request closed without merging**
